plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TransactionLogAccess.java

```diff
@@ -461,50 +478,62 @@ else if (deltaLakeTransactionLogEntry.getRemove() != null) {
         return Stream.concat(filteredCheckpointEntries, activeJsonEntries.values().stream());
     }
 
-    public Stream<RemoveFileEntry> getRemoveEntries(ConnectorSession session, TableSnapshot tableSnapshot)
-    {
-        return getEntries(
-                session,
-                tableSnapshot,
-                REMOVE,
-                entryStream -> entryStream.map(DeltaLakeTransactionLogEntry::getRemove).filter(Objects::nonNull),
-                fileSystemFactory.create(session),
-                fileFormatDataSourceStats);
-    }
-
-    public Map<Class<?>, Object> getTransactionLogEntries(
-            ConnectorSession session,
-            TableSnapshot tableSnapshot,
-            Set<CheckpointEntryIterator.EntryType> entryTypes,
-            Function<Stream<DeltaLakeTransactionLogEntry>, Stream<Object>> entryMapper)
+    public MetadataAndProtocolEntries getMetadataAndProtocolEntry(ConnectorSession session, TableSnapshot tableSnapshot)
     {
-        try (Stream<Object> entries = getEntries(
-                session,
-                tableSnapshot,
-                entryTypes,
-                (checkpointStream, jsonStream) -> entryMapper.apply(Stream.concat(checkpointStream, jsonStream.stream().map(Transaction::transactionEntries).flatMap(Collection::stream))),
-                fileSystemFactory.create(session),
-                fileFormatDataSourceStats)) {
-            return entries.collect(toImmutableMap(Object::getClass, Function.identity(), (first, second) -> second));
+        if (tableSnapshot.getCachedMetadata().isEmpty() || tableSnapshot.getCachedProtocol().isEmpty()) {
```

Comment:
There's room for improvement here.
If we're looking only for metadata & protocol, we should be reading backwards from the highest transaction log towards the checkpoint.

In case you take this request, please also test your  algorithm via `TestDeltaLakeFileOperations` as well to ensure that e.g., the checkpoint is not read when doing some table changes in the subsequent transaction logs for a `SHOW CREATE TABLE` statement.

Reply:
Yes, that would be better, but I'd like to limit scope of this PR. The same improvement could apply to getMetadataEntry and getProtocolEntry as well. You can file a GH issue about it for tracking.

Reply:
https://github.com/trinodb/trino/issues/24570